File: single-number.cpp
class Solution {
public:
#define M2
#ifdef M1
    int singleNumber(int A[], int n) {
        int x = 0; //注意：0与任何整数异或都得到该整数本身。
        for(int i=0; i<n; ++i) {
            x ^= A[i];
        }
        return x;
    }
#endif
#ifdef M2 // common solution for single number
    int singleNumber(int A[], int n) {
        const int W = sizeof(int) * 8;
        vector<int> b(W, 0);

        for(int i=0; i<n; ++i) {
            for(int j=0; j<W; ++j) {
                b[j] += 0x1 & (A[i] >> j);
                b[j] %= 2;
            }
        }

        int result = 0;
        for(int i=0; i<W; ++i) {
            result += b[i] << i;
        }

        return result;
    }
#endif
};

File: java-versions/single-number.java
/*
136. Single Number
Easy

5324

181

Add to List

Share
Given a non-empty array of integers nums, every element appears twice except for one. Find that single one.

Follow up: Could you implement a solution with a linear runtime complexity and without using extra memory?

 

Example 1:

Input: nums = [2,2,1]
Output: 1
Example 2:

Input: nums = [4,1,2,1,2]
Output: 4
Example 3:

Input: nums = [1]
Output: 1
*/

class Solution {
    // use HashMap, Time: O(N), Space: O(N)
    public int singleNumber1(int[] nums) {
        Map<Integer, Integer> m = new HashMap<>();
        for(int i : nums) {
            m.put(i, m.getOrDefault(i, 0)+1);
        }
        for(int i: nums) {
            if (m.get(i) == 1) {
                return i;
            }
        }
        throw new RuntimeException("bad request");
    }
    
    // use HashSet, Time: O(N), Space: O(N)
    public int singleNumber2(int[] nums) {
        Set<Integer> s = new HashSet<>();
        Long arraySum = 0L;
        for(int i : nums) {
            arraySum += i;
            s.add(i);
        }
        Long setSum = 0L;
        for(int i: s) {
            setSum += i;
        }
        return (int)(2 * setSum - arraySum);
    }
    
    // bit xor, Time: O(N), Space: O(1)
    public int singleNumber3(int[] nums) {
        int result = 0;
        for(int i : nums) {
            result ^= i;
        }
        return result;
    }
    
    public int singleNumber(int[] nums) {
        return singleNumber3(nums);
    }
}

